recipes/openai_ab.py

```diff
@@ -0,0 +1,365 @@
+                    self.display.render(**input_.prompt_args),
+                    responses,
+                    randomize=self.randomize,
+                )
+
+    def progress(self) -> float:
+        return 0.0
+
+    def on_exit(self, ctrl):
+        examples = ctrl.db.get_dataset_examples(ctrl.dataset)
+        counts = Counter()
+        # Get last example per ID
+        for eg in examples:
+            selected = eg.get("accept", [])
+            if not selected or len(selected) != 1 or eg["answer"] != "accept":
+                continue
+            counts[selected[0]] += 1
+        print("")
+        if not counts:
+            msg.warn("No answers found", exits=0)
+        msg.divider("Evaluation results", icon="emoji")
+        pref, _ = counts.most_common(1)[0]
+        msg.good(f"You preferred {pref}")
+        rows = [(name, count) for name, count in counts.most_common()]
+        msg.table(rows, aligns=("l", "r"))
+
+    def _get_response_batch(self, inputs: List[PromptInput]) -> List[Dict[str, str]]:
+        name1, name2 = self._choose_rivals()
+        prompts = []
+        for input_ in inputs:
+            prompts.append(self._get_prompt(name1, input_.prompt_args))
+            prompts.append(self._get_prompt(name2, input_.prompt_args))
+        responses = self._get_responses(prompts)
+        assert len(responses) == len(inputs) * 2
+        output = []
+        # Pair out the responses. There's a fancy
+        # zip way to do this but I think that's less
+        # readable
+        for i in range(0, len(responses), 2):
+            output.append({name1: responses[i], name2: responses[i + 1]})
+        return output
+
+    def _choose_rivals(self) -> Tuple[str, str]:
+        assert len(self.prompts) == 2
+        return tuple(sorted(self.prompts.keys()))
+
+    def _get_prompt(self, name: str, args: Dict) -> str:
+        return self.prompts[name].render(**args)
+
+    def _get_responses(self, prompts: List[str]) -> List[str]:
+        headers = {
+            "Authorization": f"Bearer {self.openai_api_key}",
+            "OpenAI-Organization": self.openai_api_org,
+            "Content-Type": "application/json",
+        }
+        r = _retry429(
+            lambda: httpx.post(
+                "https://api.openai.com/v1/completions",
+                headers=headers,
+                json={
+                    "model": self.model,
+                    "prompt": prompts,
+                    "temperature": self.openai_temperature,
+                    "max_tokens": self.openai_max_tokens,
+                },
+                timeout=self.openai_timeout_s,
+            ),
+            n=self.openai_n,
+            timeout_s=self.openai_timeout_s,
+        )
+        r.raise_for_status()
+        responses = r.json()
+        return [responses["choices"][i]["text"] for i in range(len(prompts))]
+
+    def _make_example(
+        self, id: str, input: str, responses: Dict[str, str], randomize: bool
```

Comment:
Should we avoid using `id` and `input` as var names, as they clash with builtins?

Reply:
I agree. Calling them `input_str` and `id_str` for now.